### PR TITLE
Work package search using embedded table

### PR DESCRIPTION
**Pull request by wielinde**
https://community.openproject.com/wp/29388

- [x] Old rails search always searches with _AND_. Never with _OR_.
- [x] Reduce search input forms to the one in the header.
- [x] Always show search input field on Desktop to make it more prominent.
- [x] Remove scope "My projects" as it is not mappable to a work package filter.
- [x] Allow old rails search to find attachment contents and file names.
- [x] Distribute search results in tabs with one tab per searchable module
- [x] Use an embedded table for displaying work packages in the Work Packages tab.
- [x] Rename tabs for modules "Wiki pages" and "Messages" to "Wiki" and "Forums"
- [x] Make tabs scrollable when screen smaller than needed to show all tabs at once
- [x] Leave the current projects scope when searching in all projects.
- [x] Make project scope selectable from within the autocompleter

### CODE REVIEW

**Comment by wielinde** [comment]
Put this work on hold until we have a clearer picture on how big we want to make that work.

**Comment by wielinde** [comment]
For reference I reopen it

**Comment by HDinger** [comment]
Hey @wielinde I had a look at the functionality. Here are the things I found:

### Styling
* The input field is not centered
bildschirmfoto 2019-02-01 um 11 18 45

* The list is not limited in its height. So wenn it is long, it overflows the actual content.
bildschirmfoto 2019-02-01 um 11 23 53

* The space between the input and the result list looks good with a header with color. On a white background it looks a bit misplaced.
bildschirmfoto 2019-02-01 um 11 25 24

* The input overlaps the logo on smaller screens.
bildschirmfoto 2019-02-01 um 11 49 06

* On tablet screen sizes, the whole search field moves one row below because there is not enough space.
bildschirmfoto 2019-02-01 um 11 55 53

* I think there is one `<td>` per row missing for the settings column. This results in strange hover effects.
bildschirmfoto 2019-02-01 um 14 10 02

* In Safari the bottom border for the filter button is missing.
bildschirmfoto 2019-02-01 um 14 57 26

* On mobile Safari the tabs are highlighted when I touch them to scroll the tab bar. This highlighting stays when I move the finger up again. Thus two elements are then highlighted (the actual tab, and the one I touched to scroll)

**Comment by wielinde** [comment]
Hey @HDinger, from the things you reported, I fixed:

- [x] Global Work package search bug
- [x] ESC button click closes the autocompleter
- [x] Prevent wrapping of header on small screens
- [x] Empty table cell for empty column

I might fix but should not block the review:

- [ ] "Advanced filter" button stays visible for closing. (incl. fix of border bottom cut off in mobile Safari)
- [ ] Fix hover state of tabs in mobile Safari

I won't fix in this PR:

- [ ] Logged out search error message.
- [ ] No results message in embedded table (https://community.openproject.com/projects/openproject/work_packages/details/29490/relations).
- [ ] Advanced styling of autocompleter results (will be worth it once ng-Select is integrated).
- [ ] Positioning of autocompleter results list (will be worth it once ng-Select is integrated).

File: app/helpers/search_helper.rb [hunk 1]
```diff
@@ -85,16 +85,17 @@ def with_notes_anchor(event, tokens)
...
   end
 
-  def type_label(t)
-    l("label_#{t.singularize}_plural", default: t.to_s.humanize)
+  def attachment_fulltexts(event)
+    attachment_strings_for(:fulltext, event)
+  end
+
+  def attachment_filenames(event)
+    # TODO: attribute "filename" might not be available on fog?
```
Comment (HDinger) [inline app/helpers/search_helper.rb]: Is this a Todo that we should handle now?

File: frontend/src/app/modules/global_search/global-search-input.component.ts [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+    this.$input.autocomplete({
+      delay: 250,
+      autoFocus: true,
+      appendTo: '#top-menu',
+      classes: {
+        'ui-autocomplete': 'search-autocomplete--results'
+      },
+      position: {
+        my: 'left top+9',
```
Comment (HDinger) [inline frontend/src/app/modules/global_search/global-search-input.component.ts:132]: What is `+9`? Is this a typo or functionality?

**Review by HDinger — CHANGES_REQUESTED** [review]
Hey @wielinde 

I was finally able to reproduce the error I told you yesterday about: You have to search for a special character. Then three different errors can occur: Either  the site crashes with a 500 code (searched for `<a`), or the site crashes with a 400 code (searched for `%`) , or the search page appears but there is not table and no search string shown (searched for `#`).

Please also have a look at the codeclimate issues and the failing `search.spec`.

File: frontend/src/app/modules/global_search/global-search-input.component.ts (hunk 2 above)
Comment (wielinde) [reply frontend/src/app/modules/global_search/global-search-input.component.ts:132]: This is syntax of jQuery UI that allows you to position what they call `Widgets`. Took me quite some time to find that :-)

File: app/helpers/search_helper.rb (hunk 1 above)
Comment (wielinde) [reply app/helpers/search_helper.rb]: Checked it and removed it.

**Pull request merged**